File: tasks/redact-spans/go/solution/redact.go
package main

import (
	"os"
	"slices"
	"strings"
)

type found struct {
	Start int
	End   int
}

type attributed struct {
	Start int
	End   int
	ID    string
}

type span struct {
	Start int      `json:"start"`
	End   int      `json:"end"`
	Rules []string `json:"rules"`
}

// findSpans reports every non-overlapping occurrence, scanning left to right.
func findSpans(item rule, points []rune) []found {
	if item.Kind == "span" {
		return []found{{Start: item.Start, End: item.End}}
	}
	value := codePoints(item.Value)
	width := len(value)
	spans := []found{}
	index := 0
	for index+width <= len(points) {
		if slices.Equal(points[index:index+width], value) {
			spans = append(spans, found{Start: index, End: index + width})
			if os.Getenv("LAB_SABOTAGE") == "overlapping-literal-matches" {
				index++
			} else {
				index += width
			}
		} else {
			index++
		}
	}
	return spans
}

func keepLongEnough(spans []found, minimum int) []found {
	kept := []found{}
	for _, item := range spans {
		if item.End-item.Start >= minimum {
			kept = append(kept, item)
		}
	}
	return kept
}

func hasOverlap(spans []found) bool {
	ordered := slices.Clone(spans)
	slices.SortFunc(ordered, func(left, right found) int {
		if left.Start != right.Start {
			return left.Start - right.Start
		}
		return left.End - right.End
	})
	for index := 1; index < len(ordered); index++ {
		if ordered[index].Start < ordered[index-1].End {
			return true
		}
	}
	return false
}

// mergeSpans combines spans that overlap or touch, keeping every contributing id.
func mergeSpans(spans []attributed) []span {
	joined := os.Getenv("LAB_SABOTAGE") != "merge-drops-touching"
	ordered := slices.Clone(spans)
	slices.SortFunc(ordered, func(left, right attributed) int {
		if left.Start != right.Start {
			return left.Start - right.Start
		}
		if left.End != right.End {
			return left.End - right.End
		}
		return strings.Compare(left.ID, right.ID)
	})
	merged := []span{}
	for _, item := range ordered {
		if len(merged) > 0 {
			last := &merged[len(merged)-1]
			if item.Start < last.End || (joined && item.Start == last.End) {
				if item.End > last.End {
					last.End = item.End
				}
				if !slices.Contains(last.Rules, item.ID) {
					last.Rules = append(last.Rules, item.ID)
				}
				continue
			}
		}
		merged = append(merged, span{
			Start: item.Start,
			End:   item.End,
			Rules: []string{item.ID},
		})
	}
	for index := range merged {
		slices.Sort(merged[index].Rules)
	}
	return merged
}

func applyMask(points []rune, spans []span, mask string) string {
	masked := slices.Clone(points)
	replacement := codePoints(mask)[0]
	for _, item := range spans {
		for index := item.Start; index < item.End; index++ {
			masked[index] = replacement
		}
	}
	return string(masked)
}

File: tasks/redact-spans/typescript/solution/redact.ts
import { codePoints } from "./parse";
import type { Rule, Span } from "./parse";

export type Found = [number, number];
export type Attributed = [number, number, string];

const sabotage = process.env["LAB_SABOTAGE"] ?? "";

/** Every non-overlapping occurrence, scanning left to right. */
export function findSpans(rule: Rule, points: string[]): Found[] {
  if (rule.kind === "span") {
    return [[rule.start, rule.end]];
  }
  const value = codePoints(rule.value);
  const width = value.length;
  const spans: Found[] = [];
  let index = 0;
  while (index + width <= points.length) {
    let same = true;
    for (let offset = 0; offset < width; offset += 1) {
      if (points[index + offset] !== value[offset]) {
        same = false;
        break;
      }
    }
    if (same) {
      spans.push([index, index + width]);
      index += sabotage === "overlapping-literal-matches" ? 1 : width;
    } else {
      index += 1;
    }
  }
  return spans;
}

export function keepLongEnough(spans: Found[], minimum: number): Found[] {
  return spans.filter(([start, end]) => end - start >= minimum);
}

export function hasOverlap(spans: Attributed[]): boolean {
  const ordered = [...spans].sort(
    (left, right) => left[0] - right[0] || left[1] - right[1],
  );
  for (let index = 1; index < ordered.length; index += 1) {
    if (ordered[index]![0] < ordered[index - 1]![1]) {
      return true;
    }
  }
  return false;
}

interface Growing {
  start: number;
  end: number;
  rules: Set<string>;
}

/** Combine spans that overlap or touch, keeping every contributing id. */
export function mergeSpans(spans: Attributed[]): Span[] {
  const joined = sabotage !== "merge-drops-touching";
  const ordered = [...spans].sort(
    (left, right) =>
      left[0] - right[0] ||
      left[1] - right[1] ||
      (left[2] < right[2] ? -1 : left[2] > right[2] ? 1 : 0),
  );
  const merged: Growing[] = [];
  for (const [start, end, id] of ordered) {
    const last = merged[merged.length - 1];
    if (last && (start < last.end || (joined && start === last.end))) {
      last.end = Math.max(last.end, end);
      last.rules.add(id);
    } else {
      merged.push({ start, end, rules: new Set([id]) });
    }
  }
  return merged.map((item) => ({
    start: item.start,
    end: item.end,
    rules: [...item.rules].sort((left, right) =>
      left < right ? -1 : left > right ? 1 : 0,
    ),
  }));
}

export function applyMask(
  points: string[],
  spans: Span[],
  mask: string,
): string {
  const masked = [...points];
  for (const span of spans) {
    for (let index = span.start; index < span.end; index += 1) {
      masked[index] = mask;
    }
  }
  return masked.join("");
}
